### ISSUE TITLE
Replace Dagger 2 with Koin

### ISSUE THREAD

**Issue by stanbar** [android10/Android-CleanArchitecture-Kotlin#39]
Hello.
[Here is a fork](https://github.com/stasbar/Android-CleanArchitecture-Kotlin) where I did migration from Dagger2 to [Koin](https://github.com/InsertKoinIO/koin). 
[Take a look](https://github.com/stasbar/Android-CleanArchitecture-Kotlin/commit/430ec545eaf3ce418ca40e5fb8ad41f27187f393) how much boilerplate code I was able to get rid of.
Don't you think that it's much simpler, easier, faster and **cleaner** way ?
- No regenerate & recompile required
- No annotations required
- self described Kotlin DSL

**Comment by android10** [comment]
DI was out of scope of this, although that does not mean we can discuss about different approaches we can use. Also the fact we are using Dagger is an implementation details. :)

**Comment by crjacinro** [comment]
Hi @stasbar , Just curious on your experience in replacing Dagger2. Was it smooth ? or You did a lot of refactoring / changes when shifting the library used?

**Comment by stasbar** [comment]
@android10 You are right, it's implementation detail. But, I was curious if you chose Dagger2 purpusely, in this project, or was it 

@crjacinro It was very smooth. I didn't add any new class, interfaces or anything like that. [Here is great article](https://android.jlelse.eu/moving-from-dagger-to-koin-simplify-your-android-development-e8c61d80cddb) about migration from Dagger to KOIN.
Basically removed whole **di** package and created [one file](https://github.com/stasbar/Android-CleanArchitecture-Kotlin/blob/master/app/src/main/kotlin/com/fernandocejas/sample/core/di/ApplicationModule.kt) with modules, which is much simpler since you have everything in one place. (Main reason I don't like Dagger, is it classes complexity and high entry point). Additionally I removed all `@Inject` annotations, and changed way of injection from
```kotlin
@Inject lateinit var movieDetailsAnimator: MovieDetailsAnimator
private lateinit var movieDetailsViewModel: MovieDetailsViewModel
override fun onCreate(savedInstanceState: Bundle?) {
    ...
    appComponent.inject(this)
    movieDetailsViewModel = viewModel(viewModelFactory)
    ...
}
```
to
```kotlin
private val movieDetailsAnimator: MovieDetailsAnimator by inject()
private val movieDetailsViewModel: MovieDetailsViewModel  by viewModel()
```
**val**s always on 👍 

I'm quite experienced in KOIN and it's brother project [Kodein](https://github.com/Kodein-Framework/Kodein-DI/), so it took me only about 20min.

**Comment by Zhuinden** [comment]
Okay, but `by viewModel()` creates the ViewModel with new instance factory, no? That's typically not what you want and want to use your own ViewModelProvider.Factory

**Comment by stasbar** [comment]
@Zhuinden from the documentation

> The by viewModel() function lazily calls the ViewModel class instance and bind it to internal ViewModels factory.

And here is also implementation
```kotlin
fun <T : ViewModel> LifecycleOwner.getViewModelByClass(
    fromActivity: Boolean = false,
    clazz: KClass<T>,
    key: String? = null,
    name: String? = null,
    parameters: Parameters = emptyParameters()
): T {
    KoinFactory.apply {
        this.parameters = parameters
        this.name = name
    }
    val viewModelProvider = when {
        this is FragmentActivity -> {
            Koin.logger.log("[ViewModel] get for FragmentActivity @ $this")
            ViewModelProvider(ViewModelStores.of(this), KoinFactory)
        }
        this is Fragment -> {
            if (fromActivity) {
                Koin.logger.log("[ViewModel] get for FragmentActivity @ ${this.activity}")
                ViewModelProvider(ViewModelStores.of(this.activity), KoinFactory)
            } else {
                Koin.logger.log("[ViewModel] get for Fragment @ $this")
                ViewModelProvider(ViewModelStores.of(this), KoinFactory)
            }
        }
        else -> error("Can't get ViewModel on $this - Is not a FragmentActivity nor a Fragment")
    }
    return if (key != null) viewModelProvider.get(
        key,
        clazz.java
    ) else viewModelProvider.get(clazz.java)
}
```
where KoinFactory is singleton factory
```kotlin
object KoinFactory : ViewModelProvider.Factory
```
